File: backend/app/schemas/tags.py
from pydantic import BaseModel, Field


class TrackTagsUpdate(BaseModel):
    """Tag fields to write. Omitted fields are untouched; null clears the tag."""

    title: str | None = Field(default=None, max_length=255)
    artists: list[str] | None = None
    album: str | None = Field(default=None, max_length=255)
    album_artist: str | None = Field(default=None, max_length=255)
    genres: list[str] | None = None
    year: int | None = Field(default=None, ge=0, le=9999)
    date: str | None = Field(default=None, max_length=20)  # raw date, overrides year
    track_number: int | None = Field(default=None, ge=0, le=999)
    disc_number: int | None = Field(default=None, ge=0, le=999)
    composer: str | None = Field(default=None, max_length=255)
    comment: str | None = Field(default=None, max_length=1000)
    copyright: str | None = Field(default=None, max_length=255)
    isrc: str | None = Field(default=None, max_length=50)
    bpm: str | None = Field(default=None, max_length=10)
    conductor: str | None = Field(default=None, max_length=255)
    language: str | None = Field(default=None, max_length=50)
    publisher: str | None = Field(default=None, max_length=255)
    lyricist: str | None = Field(default=None, max_length=255)
    website: str | None = Field(default=None, max_length=500)


class TrackFileTags(BaseModel):
    """Raw tag values as stored in the audio file."""

    artists: list[str] = []
    genres: list[str] = []
    title: str | None = None
    album: str | None = None
    album_artist: str | None = None
    date: str | None = None
    track_number: str | None = None
    disc_number: str | None = None
    composer: str | None = None
    comment: str | None = None
    copyright: str | None = None
    isrc: str | None = None
    bpm: str | None = None
    conductor: str | None = None
    language: str | None = None
    publisher: str | None = None
    lyricist: str | None = None
    website: str | None = None


class BatchTagChanges(BaseModel):
    """Fields that make sense applied to many tracks at once."""

    artists: list[str] | None = None
    album: str | None = Field(default=None, max_length=255)
    album_artist: str | None = Field(default=None, max_length=255)
    genres: list[str] | None = None
    year: int | None = Field(default=None, ge=0, le=9999)


class BatchTagsRequest(BaseModel):
    track_ids: list[int] = Field(min_length=1, max_length=1000)
    changes: BatchTagChanges


class BatchTagsResult(BaseModel):
    updated: int
    errors: list[str]

File: frontend/src/api/tags.ts
import { api } from "./http";
import type { Track } from "./types";

export interface TrackTagsUpdate {
  title?: string | null;
  artists?: string[] | null;
  album?: string | null;
  album_artist?: string | null;
  genres?: string[] | null;
  year?: number | null;
  track_number?: number | null;
  disc_number?: number | null;
}

export type BatchTagChanges = Pick<
  TrackTagsUpdate,
  "artists" | "album" | "album_artist" | "genres" | "year"
>;

export interface BatchTagsResult {
  updated: number;
  errors: string[];
}

export const updateTrackTags = (trackId: number, changes: TrackTagsUpdate) =>
  api<Track>(`/tracks/${trackId}/tags`, { method: "PATCH", body: JSON.stringify(changes) });

export const batchUpdateTags = (trackIds: number[], changes: BatchTagChanges) =>
  api<BatchTagsResult>("/tracks/tags/batch", {
    method: "POST",
    body: JSON.stringify({ track_ids: trackIds, changes }),
  });

export async function uploadAlbumCover(albumId: number, file: File): Promise<void> {
  const form = new FormData();
  form.append("file", file);
  await api<void>(`/albums/${albumId}/cover`, { method: "PUT", body: form });
}
